pkg/providers/vsphere/virtualmachine/snapshot.go

```diff
@@ -172,3 +176,161 @@ func (m snapshotMap) add(parent string, tree []types.VirtualMachineSnapshotTree)
+						vmCtx.Logger.V(5).Info("Adding fileKey", "fileKey", chain.FileKey[len(chain.FileKey)-1])
+						fileKeyList = append(fileKeyList, int(chain.FileKey[len(chain.FileKey)-1]))
+					}
+				}
+			}
+		}
+	}
+
+	// Find the keySets for FCDs from layoutEx.file
+	fcdDeviceFileKeySet := getFCDDevicesFileKeySet(vmCtx, vmDevices, vmlayout.File)
+	// Remove the fileKeys that are part of the FCDs
+	fileKeyList = slices.DeleteFunc(fileKeyList, func(key int) bool {
+		return fcdDeviceFileKeySet.Has(key)
+	})
+
+	vmCtx.Logger.V(5).Info("final fileKeyList after removing FCD related fileKeys", "fileKeyList", fileKeyList)
+
+	fileKeyMap := make(map[int]types.VirtualMachineFileLayoutExFileInfo)
+	for _, file := range vmlayout.File {
+		fileKeyMap[int(file.Key)] = file
+	}
+
+	var size int64
+	for _, fileKey := range fileKeyList {
+		if file, ok := fileKeyMap[fileKey]; ok {
+			size += file.Size
+		}
+		// Assume the size is 0 if it's not found in the fileKeyMap
+	}
+
+	return size
+}
+
+// getFCDDevicesFileKeySet extracts fileKeys of layoutEx.Files that are part of the FCDs.
+// Here we use the file name to identify the FCDs because the backingObjectID
+// is not always available. and there is no other uuid available.
+func getFCDDevicesFileKeySet(vmCtx pkgctx.VirtualMachineContext, vmDevices []types.BaseVirtualDevice, vmlayoutFileInfos []types.VirtualMachineFileLayoutExFileInfo) sets.Set[int] {
+	fcdDevicesBackingFileNamePrefixSet := sets.Set[string]{}
+	for _, device := range vmDevices {
+		vd, ok := device.(*types.VirtualDisk)
+		if !ok || vd.VDiskId == nil || vd.VDiskId.Id == "" { // FCD has VDiskId
+			continue
+		}
+
+		fileName := ""
+		switch tb := vd.Backing.(type) {
+		case *types.VirtualDiskFlatVer2BackingInfo:
+			fileName = tb.FileName
+		case *types.VirtualDiskSeSparseBackingInfo:
+			fileName = tb.FileName
+		case *types.VirtualDiskRawDiskMappingVer1BackingInfo:
+			fileName = tb.FileName
+		case *types.VirtualDiskSparseVer2BackingInfo:
+			fileName = tb.FileName
+		case *types.VirtualDiskRawDiskVer2BackingInfo:
+			fileName = tb.DescriptorFileName
+		}
+		// The device backing file name might look like below:
+		// "[sharedVmfs-0] fcd/c4f94aecf2cf42ba9bb66fb6c9fc3527-000002.vmdk"
+		// While the file name in the layoutEx.file might be like below:
+		// "[sharedVmfs-0] fcd/c4f94aecf2cf42ba9bb66fb6c9fc3527-000001.vmdk"
+		// "[sharedVmfs-0] fcd/c4f94aecf2cf42ba9bb66fb6c9fc3527-000001-sesparse.vmdk"
+		// "[sharedVmfs-0] fcd/c4f94aecf2cf42ba9bb66fb6c9fc3527-flat.vmdk"
+
+		// In order to find the corresponding fcd related file name in the layoutEx.file,
+		// get the uuid from fileName and check whether the layoutEx.file.Name contains the fcd/<uuid>.
+		// If yes, add the file.Key of that layoutEx.file to the fcdDeviceFileKeySet
+
+		// Use regex matcher to find the string like fcd/[a-zA-Z0-9] from the layoutEx.file.Name.
+		re := regexp.MustCompile(`fcd\/[a-zA-Z0-9]+`)
+		matches := re.FindStringSubmatch(fileName)
+		if len(matches) > 0 {
```

Comment:
Can you check and confirm that these snapshotted disks do not have `vDiskId` set to true in the `snapshot.config.hardware.device` that will be the easiest way to tell that these are FCDs.?

Reply:
After discussion we decided to use device keys as identifier to filter files under snapshot disk chain